File: wasm/src/runtime/id.rs
// Copyright 2020-2023 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const ACTOR_ID_UNIQUE_BYTES: usize = 12;
const TASK_ID_UNIQUE_BYTES: usize = 8;

pub const UNIQUE_ID_SIZE: usize = 28;
pub const JOB_ID_SIZE: usize = 4;

pub const ACTOR_ID_SIZE: usize = JOB_ID_SIZE + ACTOR_ID_UNIQUE_BYTES;
pub const TASK_ID_SIZE: usize = ACTOR_ID_SIZE + TASK_ID_UNIQUE_BYTES;
const MAX_OBJECT_ID_INDEX: u32 = ((1 as u64) << 32 - 1) as u32;
pub const OBJECT_ID_SIZE: usize = TASK_ID_SIZE + 4;

pub const PLACEMENT_GROUP_ID_SIZE: usize = JOB_ID_SIZE + 14;

pub type FunctionID = UniqueID;
pub type ActorClassID = UniqueID;
pub type WorkerID = UniqueID;
pub type ConfigID = UniqueID;
pub type NodeID = UniqueID;

pub trait Base<T> {
    fn size(&self) -> usize;
    fn from_random() -> T;
    fn from_binary(data: &[u8]) -> T;
    fn from_hex_string(hex: &str) -> T;
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct ObjectID {
    pub id: [u8; OBJECT_ID_SIZE],
}

impl ObjectID {
    pub fn new() -> ObjectID {
        ObjectID {
            id: [0; OBJECT_ID_SIZE],
        }
    }
}

impl Base<ObjectID> for ObjectID {
    fn size(&self) -> usize {
        OBJECT_ID_SIZE
    }

    fn from_random() -> ObjectID {
        unimplemented!()
    }

    fn from_binary(data: &[u8]) -> ObjectID {
        unimplemented!()
    }

    fn from_hex_string(hex: &str) -> ObjectID {
        unimplemented!()
    }
}

pub struct UniqueID {
    pub id: [u8; UNIQUE_ID_SIZE],
}

impl UniqueID {
    pub fn new() -> UniqueID {
        UniqueID {
            id: [0; UNIQUE_ID_SIZE],
        }
    }
}

impl Base<UniqueID> for UniqueID {
    fn size(&self) -> usize {
        UNIQUE_ID_SIZE
    }

    fn from_random() -> UniqueID {
        unimplemented!()
    }

    fn from_binary(data: &[u8]) -> UniqueID {
        unimplemented!()
    }

    fn from_hex_string(hex: &str) -> UniqueID {
        unimplemented!()
    }
}

pub struct ActorID {
    pub id: [u8; ACTOR_ID_SIZE],
}

impl ActorID {
    pub fn new() -> ActorID {
        ActorID {
            id: [0; ACTOR_ID_SIZE],
        }
    }
}

impl Base<ActorID> for ActorID {
    fn size(&self) -> usize {
        ACTOR_ID_SIZE
    }

    fn from_random() -> ActorID {
        unimplemented!()
    }

    fn from_binary(data: &[u8]) -> ActorID {
        unimplemented!()
    }

    fn from_hex_string(hex: &str) -> ActorID {
        unimplemented!()
    }
}

File: go/pkg/ids/id.go
// Copyright 2025 The Ray Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//  http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package ids

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"errors"
	"hash"
	"sync"
	"unsafe"
)

const (
	UniqueIDSize                    = 28
	JobIDSize                       = 4
	ActorIDSize                     = 16
	ActorIDUniqueBytesSize          = 12
	TaskIDSize                      = 24
	TaskIDUniqueBytesSize           = 8
	ObjectIDSize                    = 28
	ObjectIDIndexSize               = 4
	PlacementGroupIDSize            = 18
	PlacementGroupIDUniqueBytesSize = 14
	LeaseIDSize                     = 32
	LeaseIDUniqueBytesSize          = 4
)

const MaxObjectIndex int64 = (1 << (ObjectIDIndexSize * 8)) - 1

func idToHex(data []byte) string {
	return hex.EncodeToString(data)
}

// decodeHexToBytes decodes into the destination array in place to avoid a
// double allocation, and uses unsafe to avoid a heap allocation for the
// read-only source string.
func decodeHexToBytes(dst []byte, hexStr string) error {
	expectedLen := len(dst)
	if len(hexStr) != 2*expectedLen {
		return errors.New("invalid hex string length")
	}
	// Use unsafe to avoid a heap allocation (read-only path).
	hexBytes := unsafe.Slice(unsafe.StringData(hexStr), len(hexStr))
	n, err := hex.Decode(dst, hexBytes)
	if err != nil {
		return err
	}
	if n != expectedLen {
		return errors.New("hex decode length mismatch")
	}
	return nil
}

func fillRandom(data []byte) {
	_, err := rand.Read(data)
	if err != nil {
		panic("failed to generate random bytes: " + err.Error())
	}
}

// hasherPool reuses sha256 hashers to reduce GC pressure on the high-throughput path.
var hasherPool = sync.Pool{
	New: func() interface{} { return sha256.New() },
}

// generateUniqueBytesInto writes the unique bytes in place, avoiding the double
// allocation that would otherwise result from a slice escaping to the heap.
func generateUniqueBytesInto(dst []byte, jobID JobID, parentTaskID TaskID,
	counter uint64, extra int64,
) {
	h := hasherPool.Get().(hash.Hash)
	defer hasherPool.Put(h)
	h.Reset()

	h.Write(jobID.data[:])

	h.Write(parentTaskID.data[:])

	// Write the counter, using a stack allocation to avoid a heap allocation.
	var counterBytes [8]byte
	binary.LittleEndian.PutUint64(counterBytes[:], counter)
	h.Write(counterBytes[:])

	// Write the optional extra value, using a stack allocation to avoid a heap allocation.
	if extra != 0 {
		var extraBytes [8]byte
		binary.LittleEndian.PutUint64(extraBytes[:], uint64(extra))
		h.Write(extraBytes[:])
	}

	// Receive the hash into a stack-allocated array.
	var hashBuf [sha256.Size]byte
	h.Sum(hashBuf[:0])
	copy(dst, hashBuf[:len(dst)])
}

// generateUniqueBytes returns the slice returned by this function, which escapes
// to the heap; prefer generateUniqueBytesInto for direct in-place writes.
func generateUniqueBytes(jobID JobID, parentTaskID TaskID,
	counter uint64, extra int64, length int,
) []byte {
	var buf [sha256.Size]byte
	generateUniqueBytesInto(buf[:length], jobID, parentTaskID, counter, extra)
	return buf[:length]
}
